src/Adapter/Order/CommandHandler/ChangeOrdersStatusHandler.php

```diff
@@ -0,0 +1,97 @@
+ * https://opensource.org/licenses/OSL-3.0
+ * If you did not receive a copy of the license and are unable to
+ * obtain it through the world-wide-web, please send an email
+ * to license@prestashop.com so we can send you a copy immediately.
+ *
+ * DISCLAIMER
+ *
+ * Do not edit or add to this file if you wish to upgrade PrestaShop to newer
+ * versions in the future. If you wish to customize PrestaShop for your
+ * needs please refer to https://www.prestashop.com for more information.
+ *
+ * @author    PrestaShop SA <contact@prestashop.com>
+ * @copyright 2007-2019 PrestaShop SA and Contributors
+ * @license   https://opensource.org/licenses/OSL-3.0 Open Software License (OSL 3.0)
+ * International Registered Trademark & Property of PrestaShop SA
+ */
+
+namespace PrestaShop\PrestaShop\Adapter\Order\CommandHandler;
+
+use Carrier;
+use Configuration;
+use Context;
+use Order;
+use OrderHistory;
+use OrderState;
+use PrestaShop\PrestaShop\Core\Domain\Order\Command\ChangeOrdersStatusCommand;
+use PrestaShop\PrestaShop\Core\Domain\Order\CommandHandler\ChangeOrdersStatusHandlerInterface;
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\OrderException;
+use PrestaShop\PrestaShop\Core\Domain\Order\Exception\OrderNotFoundException;
+use StockAvailable;
+
+/**
+ * @internal
+ */
+final class ChangeOrdersStatusHandler implements ChangeOrdersStatusHandlerInterface
+{
+    /**
+     * @param ChangeOrdersStatusCommand $command
+     */
+    public function handle(ChangeOrdersStatusCommand $command)
+    {
+        $orderState = new OrderState($command->getNewOrderStatusId());
+
+        if ($orderState->id !== $command->getNewOrderStatusId()) {
+            throw new OrderException(
+                sprintf('Order state with ID "%s" was not found.', $command->getNewOrderStatusId())
+            );
+        }
+
+        foreach ($command->getOrderIds() as $orderId) {
+            $order = new Order($orderId->getValue());
+
+            if ($order->id !== $orderId->getValue()) {
+                throw new OrderNotFoundException($orderId);
+            }
+
+            $currentOrderState = $order->getCurrentOrderState();
+
+            if ($currentOrderState->id === $orderState->id) {
+                continue;
+            }
+
+            $history = new OrderHistory();
+            $history->id_order = $order->id;
+            $history->id_employee = (int) Context::getContext()->employee->id;
+
+            $useExistingPayment = !$order->hasInvoice();
+            $history->changeIdOrderState((int) $orderState->id, $order, $useExistingPayment);
+
+            $carrier = new Carrier($order->id_carrier, $order->id_lang);
+            $templateVars = [];
+
+            if ($history->id_order_state == Configuration::get('PS_OS_SHIPPING') && $order->shipping_number) {
+                $templateVars['{followup}'] = str_replace('@', $order->shipping_number, $carrier->url);
+            }
+
+            if ($history->addWithemail(true, $templateVars)) {
```

Comment:
@matks this one is very tricky, `$history->addWithemail()`can return `false` in two cases:

1. When `OrderHistory` was not created (which is OK)
2. When `OrderHistory` is created, but email was not sent to customer

I could workaround this by splitting this method into:

```php
if (!$history->add()) {
    // throw saving order exception
}

if (!$history->sendEmail()) {
    // throw email sending exception
}
```
Unfortunately, it does not solve issue completely. In legacy, if `$history->addWithemail()` fails, it continues to loop through other orders attempting to change their status and in the end if shows list of errors:

![Screenshot (1)](https://user-images.githubusercontent.com/15143151/58637225-ba448b80-82fa-11e9-9d56-1a84756ddeab.png)

I'm wondering how should we keep the same behavior when using command to change order status? Should we keep track of errors and throw exception with multiple errors? Wdyt?

Reply:
The question can be converted in a more global matter: "When performing one bulk action, when one of the actions fail, should we 1) stop there and display the error message to user or should we 2) skip it, attempt to perform the action on the remaining items, and at the end display a list of the failed usecases ?"

So far we have followed the strategy 1) and I like this because one error might be a hint of something wrong happening, so it's good to wait for user to re-try to make sure the bulk action was not triggered by mistake.

However the decision is a Product Choice so ping @colinegin @marionf

Reply:
Should we consider email sending failure an error? This happens pretty often, how do we inform user about it?

Reply:
@MatShir this is related to orders listing, so I let you decide with @marionf ! ;)

Reply:
@matks I think it's better to skip it and perform action in the remaining items, and display the list of the failed usecases at the end
This way if some of remaining items are not failing, the action is done for these items

Reply:
So it's decided 😉ping @sarjon 
And I'll create an issue to apply this strategy to all pages

Reply:
https://github.com/PrestaShop/PrestaShop/issues/14179